cli/crates/cli/src/output/report.rs

```diff
@@ -140,16 +143,68 @@ pub fn complete_udf_build(udf_kind: UdfKind, udf_name: &str, duration: std::time
     );
 }
 
-pub fn udf_message(udf_kind: UdfKind, udf_name: &str, message: &str, level: UdfMessageLevel) {
-    match level {
-        UdfMessageLevel::Debug => watercolor::output!("[{udf_kind} '{udf_name}'] {message}", @BrightBlack),
-        UdfMessageLevel::Error => watercolor::output!("[{udf_kind} '{udf_name}'] {message}", @Red),
-        UdfMessageLevel::Info => watercolor::output!("[{udf_kind} '{udf_name}'] {message}", @Cyan),
-        UdfMessageLevel::Warn => watercolor::output!("[{udf_kind} '{udf_name}'] {message}", @Yellow),
+pub fn udf_message(
+    udf_kind: UdfKind,
+    udf_name: &str,
+    message: &str,
+    message_level: LogLevel,
+    log_level_filter: Option<LogLevel>,
+) {
+    let Some(log_level_filter) = log_level_filter else {
+        return;
+    };
+    if message_level > log_level_filter {
+        return;
     }
+
+    let colour = match message_level {
+        LogLevel::Debug => watercolor::colored::Color::BrightBlack,
+        LogLevel::Error => watercolor::colored::Color::Red,
+        LogLevel::Info => watercolor::colored::Color::Cyan,
+        LogLevel::Warn => watercolor::colored::Color::Yellow,
+    };
+    println!("{}", format!("[{udf_kind} '{udf_name}'] {message}").color(colour));
 }
 
-pub fn operation_started(_request_id: &str, _name: &Option<String>) {}
+pub fn operation_log(_request_id: &str, log_event_type: LogEventType, log_level_filter: Option<LogLevel>) {
+    let Some(log_level_filter) = log_level_filter else {
+        return;
+    };
+    if log_level_filter < LogLevel::Info {
+        return;
+    }
+
+    let (name, r#type, colour, duration) = match log_event_type {
+        LogEventType::OperationStarted { .. } => return,
+        LogEventType::OperationCompleted { name, duration, r#type } => {
+            let colour = match r#type {
+                common::types::OperationType::Query { is_introspection } => {
+                    if is_introspection && log_level_filter < LogLevel::Debug {
+                        return;
+                    }
+                    watercolor::colored::Color::Green
+                }
+                // Pink.
```

Comment:
WDYM pink?

Reply:
It used to be pink, now I changed it to green, but didn't update the comment… The danger of comments.